### PR TITLE
implement bgp message tracking

### LINKED ISSUE #101: bgp: message tracking
`mgd` should provide a means to track the messages it has sent and received. This includes

- Updates
- Notifications
- Open

 It's probably sufficient to track keepalive messages as a peer-indexed counter.

### PR DESCRIPTION
**Pull request by rcgoodfellow**
- Fixes #101 

Adds a simple in-memory queue for incoming and outgoing messages, and an API call to retrieve them. The queue size is limited to 1024 messages in each direction. This is meant as a debugging tool and is not meant to be a persistent historian.

### CODE REVIEW

File: bgp/src/session.rs [hunk 1]
```diff
@@ -743,6 +798,10 @@ impl<Cnx: BgpConnection + 'static> SessionRunner<Cnx> {
             // Our open message has been rejected with a notifiction. Fail back
             // to the connect state, dropping the TCP connection.
             FsmEvent::Message(Message::Notification(m)) => {
+                self.message_history
```
Comment (FelixMcFelix) [inline bgp/src/session.rs:801]: I think I'm happy with all the places we're updating message history. There are few enough states and message types that it's clear enough as is, and it's also not liable to change. So we probably don't need to centralise that logic, i.e., move it out of every FSM state handler and into a `fn recv(&self, current_state: &FsmState<Cnx>) -> FsmEvent { ... }`.

**Review by FelixMcFelix — COMMENTED** [review]
Thanks Ry, this looks good. I think there are just some docstring nits, and a question on the choice of HTTP method for the new endpoint.

File: bgp/src/session.rs (hunk 1 above)
Comment (rcgoodfellow) [reply bgp/src/session.rs:801]: Yeah, this is something that's crossed my mind as well. Metrics collection could potentially also benefit from some centralization.

File: mgd/src/bgp_admin.rs [hunk 2]
```diff
@@ -721,3 +721,32 @@ fn start_bgp_session<Cnx: BgpConnection>(
...
+pub struct MessageHistoryRequest {
+    asn: u32,
+}
+
+#[derive(Debug, Serialize, JsonSchema, Clone)]
+pub struct MessageHistoryResponse {
+    by_peer: HashMap<IpAddr, MessageHistory>,
+}
+
+#[endpoint { method = POST, path = "/bgp/message-history" }]
```
[the remark this answers is not in the document]
Comment (rcgoodfellow) [reply mgd/src/bgp_admin.rs]: Yep. Great catch. Fixed

File: bgp/src/messages.rs [hunk 3]
```diff
@@ -885,23 +941,44 @@ pub enum Community {
...
 #[derive(Debug, PartialEq, Eq, Clone)]
 pub struct AsPathSegment {
+    // Indicates if this segment is a part of a set or sequence.
     pub typ: AsPathType,
+    // AS numbers in the segment.
     pub value: Vec<u16>,
 }
 
-#[derive(Debug, PartialEq, Eq, Clone)]
+// A self 4-byte describing segment found in path sets and sequences.
```
[the remark this answers is not in the document]
Comment (rcgoodfellow) [reply bgp/src/messages.rs]: Yes haha. Let this be a lesson to me, no writing comments after midnight.....

File: bgp/src/messages.rs [hunk 4]
```diff
@@ -686,14 +701,31 @@ impl PathAttributeType {
 }
 
 pub mod path_attribute_flags {
+    /// Treat a path attribute as optional
     pub const OPTIONAL: u8 = 0b10000000;
+    /// Path attribute must be redistributed
     pub const TRANSITIVE: u8 = 0b01000000;
+    /// Treat path attribute as partial
     pub const PARTIAL: u8 = 0b00100000;
+    /// If set the path attribute length is encoded in two octets instead o
```
[the remark this answers is not in the document]
Comment (rcgoodfellow) [reply bgp/src/messages.rs]: Done

**Review by FelixMcFelix — APPROVED** [review]
LGTM!

**Pull request merged**